File: ChefPocket/AIService.swift
import Foundation
import SwiftUI

struct GeminiGenerationResponse: Codable {
    struct Candidate: Codable {
        struct Content: Codable {
            struct Part: Codable {
                let text: String?
            }
            let parts: [Part]?
        }
        let content: Content?
    }
    let candidates: [Candidate]?
}

class AIService: ObservableObject {
    static let shared = AIService()
    
    @Published var isExtracting: Bool = false
    @Published var statusMessage: String = ""
    
    // Persistent API Key storage across app
    @AppStorage("chefpocket_gemini_api_key") var storedApiKey: String = ""
    
    var effectiveApiKey: String {
        storedApiKey.trimmingCharacters(in: .whitespacesAndNewlines)
    }
    
    func setApiKey(_ key: String) {
        let clean = key.trimmingCharacters(in: .whitespacesAndNewlines)
        storedApiKey = clean
        UserDefaults.standard.set(clean, forKey: "chefpocket_gemini_api_key")
    }
    
    private let cooldownSeconds: TimeInterval = 2.0
    private let dailyLimit = 50
    private let lastRequestTimeKey = "chefpocket_ai_last_req_time"
    private let dailyCountKey = "chefpocket_ai_daily_count"
    private let dailyDateKey = "chefpocket_ai_daily_date"
    
    var dailyUsageCount: Int {
        checkDailyReset()
        return UserDefaults.standard.integer(forKey: dailyCountKey)
    }
    
    var remainingDailyRequests: Int {
        max(0, dailyLimit - dailyUsageCount)
    }
    
    var hasValidApiKey: Bool {
        !storedApiKey.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }
    
    private func checkDailyReset() {
        let calendar = Calendar.current
        let today = calendar.startOfDay(for: Date())
        if let savedDate = UserDefaults.standard.object(forKey: dailyDateKey) as? Date {
            if !calendar.isDate(savedDate, inSameDayAs: today) {
                UserDefaults.standard.set(0, forKey: dailyCountKey)
                UserDefaults.standard.set(today, forKey: dailyDateKey)
            }
        } else {
            UserDefaults.standard.set(today, forKey: dailyDateKey)
            UserDefaults.standard.set(0, forKey: dailyCountKey)
        }
    }
    
    private func incrementDailyCount() {
        checkDailyReset()
        let count = UserDefaults.standard.integer(forKey: dailyCountKey)
        UserDefaults.standard.set(count + 1, forKey: dailyCountKey)
    }
    
    // MARK: - Main Extraction Pipeline
    func extractRecipe(from urlString: String, userApiKey: String? = nil) async throws -> Recipe {
        await MainActor.run {
            self.isExtracting = true
            self.statusMessage = "Analyzing video link & captions..."
        }
        
        defer {
            Task { @MainActor in
                self.isExtracting = false
                self.statusMessage = ""
            }
        }
        
        let cleanURL = urlString.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !cleanURL.isEmpty else {
            throw NSError(domain: "ChefPocket", code: 400, userInfo: [NSLocalizedDescriptionKey: "Please enter a valid video link."])
        }
        
        // 1. Resolve API key
        var activeApiKey = (userApiKey?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false) ? userApiKey!.trimmingCharacters(in: .whitespacesAndNewlines) : storedApiKey.trimmingCharacters(in: .whitespacesAndNewlines)
        
        // Check if API key is present
        guard !activeApiKey.isEmpty else {
            throw NSError(domain: "ChefPocket", code: 401, userInfo: [NSLocalizedDescriptionKey: "Gemini API Key missing. Please configure your free API Key from aistudio.google.com to extract precise recipes."])
        }
        
        // 2. Fetch video metadata (Title + Description/Caption)
        let (extractedTitle, extractedDescription, isShort) = await fetchVideoMetadataAndCaption(cleanURL)
        
        // 3. Check cooldown
        if let lastReq = UserDefaults.standard.object(forKey: lastRequestTimeKey) as? Date,
           Date().timeIntervalSince(lastReq) < cooldownSeconds {
            try? await Task.sleep(nanoseconds: UInt64(cooldownSeconds * 1_000_000_000))
        }
        
        await MainActor.run {
            self.statusMessage = "Extracting ingredients & whistle count with Gemini AI... ✨"
        }
        
        // 4. Call Gemini REST API
        let aiRecipe = try await callGeminiAPI(
            videoTitle: extractedTitle,
            videoDescription: extractedDescription,
            videoURL: cleanURL,
            apiKey: activeApiKey
        )
        
        UserDefaults.standard.set(Date(), forKey: lastRequestTimeKey)
        incrementDailyCount()
        return aiRecipe
    }
    
    // MARK: - Gemini REST API Caller with Multi-Model Fallback & Auto-Discovery
    private func callGeminiAPI(videoTitle: String, videoDescription: String, videoURL: String, apiKey: String) async throws -> Recipe {
        let cleanKey = apiKey.trimmingCharacters(in: .whitespacesAndNewlines)
        
        // Priority order of modern Gemini models
        let candidateModels = [
            "gemini-2.0-flash",
            "gemini-2.5-flash",
            "gemini-1.5-flash-latest",
            "gemini-1.5-flash",
            "gemini-2.0-flash-lite",
            "gemini-1.5-pro",
            "gemini-pro"
        ]
        
        var lastError: Error? = nil
        
        // 1. Try prioritized list of known models
        for model in candidateModels {
            do {
                return try await executeGeminiRequest(
                    model: model,
                    videoTitle: videoTitle,
                    videoDescription: videoDescription,
                    videoURL: videoURL,
                    apiKey: cleanKey
                )
            } catch let err as NSError where err.code == 404 {
                print("Gemini model \(model) returned 404. Trying next model...")
                lastError = err
                continue
            } catch {
                // If error is 400 (bad key), 429 (quota), etc., do not retry other models, return clear error
                throw error
            }
        }
        
        // 2. If all preconfigured models returned 404, query available models dynamically
        if let dynamicModel = await fetchFirstAvailableGenerateContentModel(apiKey: cleanKey) {
            print("Found dynamic model from Google: \(dynamicModel)")
            return try await executeGeminiRequest(
                model: dynamicModel,
                videoTitle: videoTitle,
                videoDescription: videoDescription,
                videoURL: videoURL,
                apiKey: cleanKey
            )
        }
        
        throw lastError ?? NSError(domain: "ChefPocket", code: 404, userInfo: [NSLocalizedDescriptionKey: "No compatible Gemini model found for this key. Please verify model permissions in Google AI Studio."])
    }
    
    private func fetchFirstAvailableGenerateContentModel(apiKey: String) async -> String? {
        guard let url = URL(string: "https://generativelanguage.googleapis.com/v1beta/models?key=\(apiKey)") else { return nil }
        guard let (data, resp) = try? await URLSession.shared.data(from: url),
              let http = resp as? HTTPURLResponse, (200...299).contains(http.statusCode) else { return nil }
        
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let models = json["models"] as? [[String: Any]] else { return nil }
        
        for m in models {
            if let name = m["name"] as? String,
               let methods = m["supportedGenerationMethods"] as? [String],
               methods.contains("generateContent") {
                let clean = name.replacingOccurrences(of: "models/", with: "")
                if clean.contains("flash") || clean.contains("gemini") {
                    return clean
                }
            }
        }
        return nil
    }
    
    private func executeGeminiRequest(model: String, videoTitle: String, videoDescription: String, videoURL: String, apiKey: String) async throws -> Recipe {
        guard let endpoint = URL(string: "https://generativelanguage.googleapis.com/v1beta/models/\(model):generateContent?key=\(apiKey)") else {
            throw NSError(domain: "ChefPocket", code: 400, userInfo: [NSLocalizedDescriptionKey: "Invalid Gemini endpoint configuration."])
        }
        
        let promptText = """
        You are an elite Michelin-trained executive chef assistant. Analyze this video metadata and extract the exact culinary recipe.
        
        Video Title: "\(videoTitle)"
        Video Caption / Description: "\(videoDescription)"
        Video URL: "\(videoURL)"
        
        CRITICAL RULES:
        1. Identify the exact dish name from the title/caption.
        2. Extract authentic ingredients with precise metric amounts (grams, ml, tbsp, tsp, piece).
        3. If it is an Indian pressure cooker dish (dal, rajma, chana, biryani, mutton, etc.), determine the exact cooker whistle count (e.g. 2, 3, 4 whistles). If not a pressure cooker recipe, set whistleCount to null.
        4. Accurately calculate calories and protein grams per serving.
        5. Provide numbered, professional cooking instructions with specific heat levels and cooking cues. Never include YouTube URLs or 'extracted from' lines in instructions.
        6. Determine Diet ("Veg" or "Non-Veg").
        7. Determine Cuisine ("Indian Regional", "Continental & Italian", "Asian & Indo-Chinese", "Mexican & Tex-Mex", "Middle Eastern", "Cafe & Bistro", "Bakery & Breads", or "Drinks & Brews").
        8. Determine Category ("Sabzi", "Dal", "High-Protein", "Breakfast", "Street Food", "Rice & Biryani", "Bakery", "Drinks & Shakes", or "Fusion").
        
        STRICT REQUIREMENT: Respond ONLY with a valid JSON object (no markdown quotes, no explanation, no text before or after):
        {
          "title": "Dish Name",
          "category": "Sabzi",
          "cuisine": "Indian Regional",
          "diet": "Veg",
          "mealTypes": ["Lunch", "Dinner"],
          "prepTimeMinutes": 25,
          "calories": 380,
          "proteinGrams": 24,
          "whistleCount": 3,
          "tags": ["Authentic", "Video Import", "High-Protein"],
          "ingredients": [
            {"name": "Paneer / Chicken", "amount": 250, "unit": "g"},
            {"name": "Desi Ghee", "amount": 2, "unit": "tbsp"},
            {"name": "Cumin Seeds", "amount": 1, "unit": "tsp"}
          ],
          "instructions": [
            "Heat ghee in a pan on medium heat and splutter cumin seeds.",
            "Add aromatics, sauté until golden brown, and add spices.",
            "Cover and cook until tender. Serve hot."
          ]
        }
        """
        
        let requestBody: [String: Any] = [
            "contents": [
                [
                    "parts": [
                        ["text": promptText]
                    ]
                ]
            ],
            "generationConfig": [
                "temperature": 0.2
            ]
        ]
        
        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: requestBody)
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResp = response as? HTTPURLResponse else {
            throw NSError(domain: "ChefPocket", code: 500, userInfo: [NSLocalizedDescriptionKey: "No response received from Gemini server."])
        }
        
        // Handle API errors gracefully with user-friendly messages
        if !(200...299).contains(httpResp.statusCode) {
            var errMsg = "Gemini API Error (\(httpResp.statusCode))"
            if let errJSON = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
               let errObj = errJSON["error"] as? [String: Any],
               let msg = errObj["message"] as? String {
                if msg.contains("API_KEY_INVALID") || msg.contains("API key not valid") {
                    errMsg = "Invalid Gemini API Key. Please verify your key at aistudio.google.com."
                } else if msg.contains("RESOURCE_EXHAUSTED") || httpResp.statusCode == 429 {
                    errMsg = "Gemini quota exceeded. Please wait a moment or use a different key."
                } else {
                    errMsg = "\(msg) (Status: \(httpResp.statusCode))"
                }
            }
            throw NSError(domain: "GeminiAPI", code: httpResp.statusCode, userInfo: [NSLocalizedDescriptionKey: errMsg])
        }
        
        let geminiResp = try JSONDecoder().decode(GeminiGenerationResponse.self, from: data)
        guard let rawJSONText = geminiResp.candidates?.first?.content?.parts?.first?.text else {
            throw NSError(domain: "ChefPocket", code: 422, userInfo: [NSLocalizedDescriptionKey: "Could not read recipe structure from AI response."])
        }
        
        // Extract substring between first '{' and last '}'
        var cleaned = rawJSONText
            .replacingOccurrences(of: "```json", with: "")
            .replacingOccurrences(of: "```", with: "")
            .trimmingCharacters(in: .whitespacesAndNewlines)
            
        if let firstBrace = cleaned.firstIndex(of: "{"),
           let lastBrace = cleaned.lastIndex(of: "}") {
            cleaned = String(cleaned[firstBrace...lastBrace])
        }
        
        guard let jsonData = cleaned.data(using: .utf8),
              let dict = try JSONSerialization.jsonObject(with: jsonData) as? [String: Any] else {
            throw NSError(domain: "ChefPocket", code: 422, userInfo: [NSLocalizedDescriptionKey: "Failed to parse structured recipe JSON from AI."])
        }
        
        // Parse fields
        let title = dict["title"] as? String ?? (videoTitle.isEmpty ? "AI Extracted Dish" : videoTitle)
        
        let catRaw = dict["category"] as? String ?? "Sabzi"
        let category = RecipeCategory(rawValue: catRaw) ?? .sabzi
        
        let cuRaw = dict["cuisine"] as? String ?? "Indian Regional"
        let cuisine = Cuisine(rawValue: cuRaw) ?? .indian
        
        let dietRaw = dict["diet"] as? String ?? "Veg"
        let diet = DietType(rawValue: dietRaw) ?? .veg
        
        var mealTypes: [MealType] = []
        if let mealsArr = dict["mealTypes"] as? [String] {
            mealTypes = mealsArr.compactMap { MealType(rawValue: $0) }
        }
        if mealTypes.isEmpty { mealTypes = [.lunch, .dinner] }
        
        let prep = dict["prepTimeMinutes"] as? Int ?? 25
        let cals = dict["calories"] as? Int ?? 380
        let protein = dict["proteinGrams"] as? Int ?? 22
        let whistles = dict["whistleCount"] as? Int
        let tags = dict["tags"] as? [String] ?? ["AI Extracted", "Chef Verified"]
        
        var ingredients: [Ingredient] = []
        if let ingArr = dict["ingredients"] as? [[String: Any]] {
            for item in ingArr {
                let name = item["name"] as? String ?? "Ingredient"
                let amt = (item["amount"] as? NSNumber)?.doubleValue ?? 1.0
                let unit = item["unit"] as? String ?? "portion"
                ingredients.append(Ingredient(name: name, amount: amt, unit: unit))
            }
        }
        if ingredients.isEmpty {
            ingredients = [Ingredient(name: "Main ingredient", amount: 250, unit: "g")]
        }
        
        var rawInstructions = dict["instructions"] as? [String] ?? [
            "Prepare and chop all fresh ingredients as listed.",
            "Heat pan or pressure cooker on medium heat, sauté aromatics and spices.",
            "Add main ingredients and simmer until cooked thoroughly.",
            "Finish with fresh garnish and serve hot."
        ]
        let instructions = rawInstructions.filter {
            let l = $0.lowercased().trimmingCharacters(in: .whitespacesAndNewlines)
            return !l.starts(with: "extracted from") && !l.starts(with: "source:") && !l.contains("http")
        }
        
        return Recipe(
            title: title,
            category: category,
            cuisine: cuisine,
            diet: diet,
            mealTypes: mealTypes,
            isUserCreated: true,
            tags: tags,
            sourceURL: videoURL,
            prepTimeMinutes: prep,
            calories: cals,
            proteinGrams: protein,
            whistleCount: whistles,
            ingredients: ingredients,
            instructions: instructions
        )
    }
    
    // MARK: - Video Metadata & Caption Scraper
    private func fetchVideoMetadataAndCaption(_ urlString: String) async -> (title: String, description: String, isShort: Bool) {
        let isShort = urlString.contains("shorts") || urlString.contains("youtube") || urlString.contains("youtu.be")
        var title = ""
        var description = ""
        
        // 1. Normalize YouTube URL to canonical watch link for metadata fetching
        var targetURL = urlString
        if urlString.contains("shorts/") {
            let parts = urlString.components(separatedBy: "shorts/")
            if let id = parts.last?.split(separator: "?").first?.split(separator: "/").first {
                targetURL = "https://www.youtube.com/watch?v=\(id)"
            }
        } else if urlString.contains("youtu.be/") {
            let parts = urlString.components(separatedBy: "youtu.be/")
            if let id = parts.last?.split(separator: "?").first?.split(separator: "/").first {
                targetURL = "https://www.youtube.com/watch?v=\(id)"
            }
        }
        
        // 2. Query oEmbed with iOS Safari User-Agent
        if let encoded = targetURL.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed),
           let oEmbedURL = URL(string: "https://www.youtube.com/oembed?url=\(encoded)&format=json") {
            var req = URLRequest(url: oEmbedURL)
            req.setValue("Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15", forHTTPHeaderField: "User-Agent")
            if let (data, _) = try? await URLSession.shared.data(for: req),
               let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
               let fetchedTitle = json["title"] as? String, !fetchedTitle.isEmpty {
                title = fetchedTitle
            }
        }
        
        // 3. Query Webpage OpenGraph tags (og:title, og:description) to extract creator caption
        if let pageURL = URL(string: targetURL) {
            var req = URLRequest(url: pageURL)
            req.setValue("Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15", forHTTPHeaderField: "User-Agent")
            if let (data, _) = try? await URLSession.shared.data(for: req),
               let html = String(data: data, encoding: .utf8) {
                // Extract og:title if empty
                if title.isEmpty {
                    if let range = html.range(of: "<meta property=\"og:title\" content=\"") {
                        let after = html[range.upperBound...]
                        if let end = after.firstIndex(of: "\"") {
                            title = String(after[..<end])
                        }
                    }
                }
                // Extract og:description (often has ingredients & recipe steps!)
                if let range = html.range(of: "<meta property=\"og:description\" content=\"") {
                    let after = html[range.upperBound...]
                    if let end = after.firstIndex(of: "\"") {
                        description = String(after[..<end])
                    }
                } else if let range = html.range(of: "<meta name=\"description\" content=\"") {
                    let after = html[range.upperBound...]
                    if let end = after.firstIndex(of: "\"") {
                        description = String(after[..<end])
                    }
                }
            }
        }
        
        if title.isEmpty {
            title = isShort ? "Chef Special YouTube Dish" : "Trending Video Dish"
        }
        
        return (title, description, isShort)
    }
}

File: android/app/src/main/java/com/chefpocket/app/network/AIService.kt
package com.chefpocket.app.network

import com.chefpocket.app.data.models.*
import com.google.gson.Gson
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.net.URLEncoder
import java.util.concurrent.TimeUnit

class AIService {
    companion object {
        val shared = AIService()
    }

    private val client = OkHttpClient.Builder()
        .connectTimeout(20, TimeUnit.SECONDS)
        .readTimeout(45, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    private val gson = Gson()

    suspend fun extractRecipe(
        urlString: String,
        userApiKey: String?,
        preferredModel: String = "gemini-3.6-flash",
        statusCallback: (String) -> Unit = {}
    ): Recipe = withContext(Dispatchers.IO) {
        val cleanURL = urlString.trim()
        if (cleanURL.isEmpty()) {
            throw IllegalArgumentException("Please provide a valid YouTube Shorts or Instagram Reels link.")
        }

        val activeKey = userApiKey?.trim() ?: ""
        if (activeKey.isEmpty()) {
            throw IllegalArgumentException("Gemini API Key missing. Please add your free key in the setup card below.")
        }

        statusCallback("Analyzing video link & creator ingredients...")
        val (title, description) = fetchVideoMetadataAndCaption(cleanURL)

        statusCallback("Generating recipe with Gemini AI... ✨")
        callGeminiAPI(
            videoTitle = title,
            videoDescription = description,
            videoURL = cleanURL,
            apiKey = activeKey,
            preferredModel = preferredModel
        )
    }

    private fun fetchVideoMetadataAndCaption(urlString: String): Pair<String, String> {
        var targetURL = urlString
        if (urlString.contains("shorts/")) {
            val id = urlString.substringAfter("shorts/").substringBefore("?").substringBefore("/")
            targetURL = "https://www.youtube.com/watch?v=$id"
        } else if (urlString.contains("youtu.be/")) {
            val id = urlString.substringAfter("youtu.be/").substringBefore("?").substringBefore("/")
            targetURL = "https://www.youtube.com/watch?v=$id"
        }

        var title = ""
        var description = ""

        // 1. Query oEmbed with Mobile Safari User-Agent
        try {
            val encoded = URLEncoder.encode(targetURL, "UTF-8")
            val oembedUrl = "https://www.youtube.com/oembed?url=$encoded&format=json"
            val req = Request.Builder()
                .url(oembedUrl)
                .header("User-Agent", "Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko)")
                .build()

            client.newCall(req).execute().use { resp ->
                if (resp.isSuccessful) {
                    val body = resp.body?.string()
                    if (!body.isNullOrBlank()) {
                        val json = JsonParser.parseString(body).asJsonObject
                        if (json.has("title")) {
                            title = json.get("title").asString
                        }
                    }
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

        // 2. Fetch page HTML for meta description
        try {
            val pageReq = Request.Builder()
                .url(targetURL)
                .header("User-Agent", "Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko)")
                .build()

            client.newCall(pageReq).execute().use { resp ->
                if (resp.isSuccessful) {
                    val html = resp.body?.string() ?: ""
                    if (html.contains("property=\"og:description\" content=\"")) {
                        description = html.substringAfter("property=\"og:description\" content=\"").substringBefore("\"")
                    } else if (html.contains("name=\"description\" content=\"")) {
                        description = html.substringAfter("name=\"description\" content=\"").substringBefore("\"")
                    }
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

        if (title.isEmpty()) {
            title = if (urlString.contains("shorts") || urlString.contains("youtube")) "Chef Special YouTube Dish" else "Trending Video Dish"
        }
        return Pair(title, description)
    }

    private suspend fun callGeminiAPI(
        videoTitle: String,
        videoDescription: String,
        videoURL: String,
        apiKey: String,
        preferredModel: String
    ): Recipe {
        val candidateModels = mutableListOf<String>()
        if (preferredModel.isNotBlank()) {
            candidateModels.add(preferredModel.trim())
        }
        val defaults = listOf(
            "gemini-3.6-flash",
            "gemini-3.8-flash",
            "gemini-3.5-flash",
            "gemini-3.0-flash",
            "gemini-2.5-flash",
            "gemini-2.0-flash",
            "gemini-1.5-flash-latest",
            "gemini-1.5-flash",
            "gemini-2.0-flash-lite"
        )
        for (m in defaults) {
            if (!candidateModels.contains(m)) {
                candidateModels.add(m)
            }
        }

        var lastException: Exception? = null

        for (model in candidateModels) {
            try {
                return executeGeminiRequest(model, videoTitle, videoDescription, videoURL, apiKey)
            } catch (e: GeminiNotFoundException) {
                lastException = e
                continue
            } catch (e: Exception) {
                throw e
            }
        }

        // Auto-Discovery fallback
        val dynamicModel = fetchFirstAvailableModel(apiKey)
        if (dynamicModel != null) {
            return executeGeminiRequest(dynamicModel, videoTitle, videoDescription, videoURL, apiKey)
        }

        throw lastException ?: Exception("Could not find a supported Gemini model. Verify key permissions in Google AI Studio.")
    }

    private fun fetchFirstAvailableModel(apiKey: String): String? {
        return try {
            val url = "https://generativelanguage.googleapis.com/v1beta/models?key=$apiKey"
            val req = Request.Builder().url(url).build()
            client.newCall(req).execute().use { resp ->
                if (resp.isSuccessful) {
                    val body = resp.body?.string() ?: return null
                    val json = JsonParser.parseString(body).asJsonObject
                    if (json.has("models")) {
                        val arr = json.getAsJsonArray("models")
                        for (item in arr) {
                            val obj = item.asJsonObject
                            val name = obj.get("name")?.asString ?: ""
                            val methods = obj.getAsJsonArray("supportedGenerationMethods")?.map { it.asString } ?: emptyList()
                            if (methods.contains("generateContent")) {
                                val clean = name.replace("models/", "")
                                if (clean.contains("flash") || clean.contains("gemini")) {
                                    return clean
                                }
                            }
                        }
                    }
                }
                null
            }
        } catch (e: Exception) {
            null
        }
    }

    private fun executeGeminiRequest(
        model: String,
        videoTitle: String,
        videoDescription: String,
        videoURL: String,
        apiKey: String
    ): Recipe {
        val endpoint = "https://generativelanguage.googleapis.com/v1beta/models/$model:generateContent?key=$apiKey"

        val promptText = """
        You are an elite Michelin-trained executive chef assistant. Analyze this video metadata and extract the exact culinary recipe.
        
        Video Title: "$videoTitle"
        Video Caption / Description: "$videoDescription"
        Video URL: "$videoURL"
        
        CRITICAL RULES:
        1. Identify the exact dish name from the title/caption.
        2. Extract authentic ingredients with precise metric amounts (grams, ml, tbsp, tsp, piece).
        3. If it is an Indian pressure cooker dish (dal, rajma, chana, biryani, mutton, etc.), determine the exact cooker whistle count (e.g. 2, 3, 4 whistles). If not a pressure cooker recipe, set whistleCount to null.
        4. Accurately calculate calories and protein grams per serving.
        5. Provide numbered, professional cooking instructions with specific heat levels and cooking cues. Never include YouTube URLs or 'extracted from' lines in instructions.
        6. Determine Diet ("Veg" or "Non-Veg").
        7. Determine Cuisine ("Indian Regional", "Continental & Italian", "Asian & Indo-Chinese", "Mexican & Tex-Mex", "Middle Eastern", "Cafe & Bistro", "Bakery & Breads", or "Drinks & Brews").
        8. Determine Category ("Sabzi", "Dal", "High-Protein", "Breakfast", "Street Food", "Rice & Biryani", "Bakery", "Drinks & Shakes", or "Fusion").
        
        STRICT REQUIREMENT: Respond ONLY with a valid JSON object (no markdown quotes, no explanation, no text before or after):
        {
          "title": "Dish Name",
          "category": "Sabzi",
          "cuisine": "Indian Regional",
          "diet": "Veg",
          "mealTypes": ["Lunch", "Dinner"],
          "prepTimeMinutes": 25,
          "calories": 380,
          "proteinGrams": 24,
          "whistleCount": 3,
          "tags": ["Authentic", "Video Import", "High-Protein"],
          "ingredients": [
            {"name": "Paneer / Chicken", "amount": 250, "unit": "g"},
            {"name": "Desi Ghee", "amount": 2, "unit": "tbsp"},
            {"name": "Cumin Seeds", "amount": 1, "unit": "tsp"}
          ],
          "instructions": [
            "Heat ghee in a pan on medium heat and splutter cumin seeds.",
            "Add aromatics, sauté until golden brown, and add spices.",
            "Cover and cook until tender. Serve hot."
          ]
        }
        """.trimIndent()

        val rootJson = JsonObject()
        val contentsArr = com.google.gson.JsonArray()
        val contentObj = JsonObject()
        val partsArr = com.google.gson.JsonArray()
        val partObj = JsonObject()
        partObj.addProperty("text", promptText)
        partsArr.add(partObj)
        contentObj.add("parts", partsArr)
        contentsArr.add(contentObj)
        rootJson.add("contents", contentsArr)

        val genConfig = JsonObject()
        genConfig.addProperty("temperature", 0.2)
        rootJson.add("generationConfig", genConfig)

        val body = rootJson.toString().toRequestBody("application/json".toMediaType())
        val req = Request.Builder()
            .url(endpoint)
            .post(body)
            .build()

        client.newCall(req).execute().use { resp ->
            val respCode = resp.code
            val respBody = resp.body?.string() ?: ""

            if (!resp.isSuccessful) {
                if (respCode == 404) {
                    throw GeminiNotFoundException("Model $model returned 404")
                }
                var errMsg = "Gemini API Error ($respCode)"
                try {
                    val errJson = JsonParser.parseString(respBody).asJsonObject
                    if (errJson.has("error")) {
                        val errObj = errJson.getAsJsonObject("error")
                        val msg = errObj.get("message")?.asString ?: ""
                        if (msg.contains("API_KEY_INVALID") || msg.contains("API key not valid")) {
                            errMsg = "Invalid Gemini API Key. Please verify in Google AI Studio."
                        } else if (msg.contains("RESOURCE_EXHAUSTED") || respCode == 429) {
                            errMsg = "Gemini quota exceeded. Please wait a moment."
                        } else {
                            errMsg = "$msg ($respCode)"
                        }
                    }
                } catch (e: Exception) {
                    // fallback
                }
                throw Exception(errMsg)
            }

            // Parse response
            val root = JsonParser.parseString(respBody).asJsonObject
            val candidateText = root.getAsJsonArray("candidates")
                ?.get(0)?.asJsonObject
                ?.getAsJsonObject("content")
                ?.getAsJsonArray("parts")
                ?.get(0)?.asJsonObject
                ?.get("text")?.asString
                ?: throw Exception("Could not parse recipe structure from AI response.")

            var cleanJson = candidateText.replace("```json", "").replace("```", "").trim()
            if (cleanJson.contains("{") && cleanJson.contains("}")) {
                cleanJson = cleanJson.substring(cleanJson.indexOf("{"), cleanJson.lastIndexOf("}") + 1)
            }

            val recipeObj = JsonParser.parseString(cleanJson).asJsonObject
            val title = recipeObj.get("title")?.asString ?: videoTitle
            val category = recipeObj.get("category")?.asString ?: "Sabzi"
            val cuisine = recipeObj.get("cuisine")?.asString ?: "Indian Regional"
            val diet = recipeObj.get("diet")?.asString ?: "Veg"
            val prep = recipeObj.get("prepTimeMinutes")?.asInt ?: 20
            val cals = recipeObj.get("calories")?.asInt ?: 350
            val protein = recipeObj.get("proteinGrams")?.asInt ?: 20
            val whistle = if (recipeObj.has("whistleCount") && !recipeObj.get("whistleCount").isJsonNull) {
                recipeObj.get("whistleCount").asInt
            } else null

            val ings = mutableListOf<Ingredient>()
            recipeObj.getAsJsonArray("ingredients")?.forEach { item ->
                val o = item.asJsonObject
                ings.add(
                    Ingredient(
                        name = o.get("name")?.asString ?: "Ingredient",
                        amount = o.get("amount")?.asDouble ?: 1.0,
                        unit = o.get("unit")?.asString ?: "g"
                    )
                )
            }

            val steps = mutableListOf<String>()
            recipeObj.getAsJsonArray("instructions")?.forEach { step ->
                val s = step.asString
                val lower = s.lowercase().trim()
                if (!lower.startsWith("extracted from") && !lower.startsWith("source:") && !lower.contains("http")) {
                    steps.add(s)
                }
            }

            return Recipe(
                title = title,
                category = category,
                cuisine = cuisine,
                diet = diet,
                isUserCreated = true,
                sourceURL = videoURL,
                prepTimeMinutes = prep,
                calories = cals,
                proteinGrams = protein,
                whistleCount = whistle,
                ingredients = if (ings.isNotEmpty()) ings else listOf(Ingredient(name = "Main ingredients", amount = 250.0, unit = "g")),
                instructions = if (steps.isNotEmpty()) steps else listOf("Prepare ingredients according to video instructions, cook and serve hot.")
            )
        }
    }
}

class GeminiNotFoundException(msg: String) : Exception(msg)
